content/en/blog/_posts/2023-03-08-memory-qos-cgroups-v2/index.md

```diff
@@ -0,0 +1,260 @@
+
+**Note**: If a container has no memory limit specified, we substitute limits.memory for node allocatable memory.
+
+### Support for Pod QoS classes
+Based on user feedback for the Alpha feature in Kubernetes v1.22, some users would like 
+to opt out of MemoryQoS on a per-pod basis to ensure there is no early memory throttling.
+Therefore, in Kubernetes v1.27 Memory QOS also supports memory.high to be set as per 
+Quality of Service(QoS) for Pod classes. Following are the different cases for memory.high
+as per QOS classes:
+
+1. **Guaranteed pods** by their QoS definition require memory requests=memory limits and are 
+not overcommitted. Hence MemoryQoS feature is disabled on those pods by not setting 
+memory.high. This ensures that Guaranteed pods can fully use their memory requests up
+to their set limit, and not hit any throttling.
+
+2. **Burstable pods** by their QoS definition require at least one container in the Pod with
+CPU or memory request or limit set.
+    
+    * When requests.memory and limits.memory are set, the formula is used as-is:
+    ![](./container-memory-high-limit.svg)  
+    <sub>i: the i<sup>th</sup> container in one pod</sub>
+
+    * When requests.memory is set, limits.memory is not set, we substitute limits.memory
+    for node allocatable memory in the formula:
+    ![](./container-memory-high-no-limits.svg)  
+    <sub>i: the i<sup>th</sup> container in one pod</sub>
+
+3. BestEffort The pod gets a BestEffort class if limits.memory and requests.memory are not 
+set. We set requests.memory = 0 and substitute limits.memory for node allocatable memory 
+in the formula:
+
+    ![](./container-memory-high-best-effort.svg)  
+    <sub>i: the i<sup>th</sup> container in one pod</sub>
+
+### Cgroup Hierarchy
+When container memory requests are made, kubelet passes `memory.min` to the back-end 
+CRI runtime (possibly containerd, cri-o) via the `Unified` field in CRI during 
+container creation. The `memory.min` in container level cgroup will be set to:  
+
+![](./container-memory-min.svg)  
+<sub>i: the i<sup>th</sup> container in one pod</sub>
+
+Since the `memory.min` interface requires that the ancestor cgroup directories are all 
+set, the pod and node cgroup directories need to be set correctly. 
+
+`memory.min` in pod level cgroup:  
+![](./pod-memory-min.svg)  
+<sub>i: the i<sup>th</sup> container in one pod</sub>
+
+`memory.min` in node level cgroup:  
+![](./node-memory-min.svg)  
+<sub>i: the i<sup>th</sup> pod in one node, j: the j<sup>th</sup> container in 
+one pod</sub>
+
+Kubelet will manage the cgroup hierarchy of the pod level and node level cgroups 
+directly using runc libcontainer library, while container cgroup limits are managed by the container runtime.
+
+## How do I use it?
+Here are the prerequisites for enabling Memory QoS on your Linux node, some of these are 
+related to [Kubernetes support for cgroup v2](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2254-cgroup-v2).
```

Comment:
It might be worth to link directly to the cgroupv2 k8s docs - https://kubernetes.io/docs/concepts/architecture/cgroups/ instead of the KEP